In [28]:
#Importing all the Necessary Libraries
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
from pathlib import Path

import warnings
warnings.simplefilter("ignore", UserWarning)


## 1. What proportion of patients experienced readmission and mortality within 28 days, 3 months, and 6 months of discharge?
### Reasoning:  

Hospital readmissions and post-discharge mortality are key quality-of-care metrics.

Patients  28 days,3 months, and 6 months readmission rates are standard quality metrics hospitals report (e.g., CMS 30-day readmission measures use a similar window), so these numbers let you compare against national or peer benchmarks.Tracking these three windows over time or across departments highlights whether early (28-day, often "preventable") readmissions vs. later ones (3-6 month, often disease-progression driven) are the bigger problem.

Reporting mortality across the same three windows lets us compare this cohort against national or peer benchmarks using a consistent, internationally recognized time frame, rather than an arbitrary cutoff

### Metrics used:
The raw dataset gives us two ways to measure them:

<h6>Flag columns</h6> (readmission_28d, readmission_3m, readmission_6m, death_28d, etc.) — pre-computed binary indicators
<h6>Time-to-event columns</h6> (readmission_time_days, death_time_days) — the actual number of days until the event occurred

We used second method to measure readmission rate and mortality rate for accuracy instead of flag columns.

admission_date – shows when the patient was admitted

readmission_time_days -the actual number of days to readmit hospital

death_time_days -the actual number of days until the death


Reading the hospitalization_discharged file

In [29]:
#Reading the file 
df_hospitalization = pd.read_csv("Team05_PyCraft_Cleaned_data/hospitalization_discharge_cleaned.csv")

In [30]:
df_hospitalization['admission_date'] = pd.to_datetime(df_hospitalization['admission_date'])
   
# Sort by patient and admission chronological order
df_hospitalization = df_hospitalization.sort_values(by=['patient_id', 'admission_date'])

# ============================================================
# Patients Readmission rate in 28 days,3 months, 6 months
# ============================================================
valid_discharges = df_hospitalization[df_hospitalization['readmission_time_days'] >= 0]
total_readmitted_patients = valid_discharges['patient_id'].nunique()

total_patients = df_hospitalization['patient_id'].nunique()


# Calculate proportions
rate_28d = (valid_discharges['readmission_time_days'] <= 28).sum() / total_patients
rate_3m = (valid_discharges['readmission_time_days'] <= 90).sum() / total_patients
rate_6m = (valid_discharges['readmission_time_days'] <= 180).sum() / total_patients


print("=== Readmission rate===")
print(f"Total patients:{total_patients}")
print(f"Total readmitted patients:{total_readmitted_patients}")
print(f"28-Day Readmission Rate: {rate_28d:.2%}")
print(f"3-Month Readmission Rate: {rate_3m:.2%}")
print(f"6-Month Readmission Rate: {rate_6m:.2%}")
print()
# ============================================================
# Patients Mortality rate in 28 days,3 months, 6 months
# ============================================================
valid_deaths = df_hospitalization[df_hospitalization['death_time_days'] >= 0]
total_died_patients = valid_deaths['patient_id'].nunique()

total_patients = df_hospitalization['patient_id'].nunique()


# Calculate proportions
mortality_28d = (valid_deaths['death_time_days'] <= 28).sum() / total_patients
mortality_3m = (valid_deaths['death_time_days'] <= 90).sum() / total_patients
mortality_6m = (valid_deaths['death_time_days'] <= 180).sum() / total_patients

print("=== Mortaility Rate ===")
print(f"Total patients:{total_patients}")
print(f"Total died patients:{total_died_patients}")
print(f"28-Day Mortality Rate: {mortality_28d:.2%}")
print(f"3-Month Mortality Rate: {mortality_3m:.2%}")
print(f"6-Month Mortality Rate: {mortality_6m:.2%}")
print()

=== Readmission rate===
Total patients:2008
Total readmitted patients:901
28-Day Readmission Rate: 6.97%
3-Month Readmission Rate: 24.80%
6-Month Readmission Rate: 34.96%

=== Mortaility Rate ===
Total patients:2008
Total died patients:44
28-Day Mortality Rate: 1.84%
3-Month Mortality Rate: 2.04%
6-Month Mortality Rate: 2.04%



### Insights:

Total patients:   2008<br/>
Total readmitted patients: 901<br/>
Total died patients:  44<br/>

28-Day Readmission Rate: 6.97% ,
3-Month Readmission Rate: 24.80% ,
6-Month Readmission Rate: 34.96%

28-Day Mortality Rate: 1.84% ,
3-Month Mortality Rate: 2.04% ,
6-Month Mortality Rate: 2.04%


1. Readmission is a far bigger problem than mortality in this cohort</li>
2. The two outcomes grow at very different rates over time
3. Mortality risk resolves early; readmission risk does not
    
Mortality-focused interventions (early follow-up, close monitoring) should be front-loaded into the first 1–3 months.<br/>
Readmission-focused interventions need to stay active across the full 6-month horizon, since new risk keeps emerging even in months 4–6.